# Compare Toronto and New York 

## Poject goal: To find the similarities between Toronto and New York

## Table of contents
* [Introduction: Business Problem](#introduction)
* [Data](#data)
* [Methodology](#methodology)
* [Analysis](#analysis)
* [Results and Discussion](#results)
* [Conclusion](#conclusion)



## Introduction: Business Problem <a name="introduction"></a>

New York City and the city of Toronto are very diverse and are the financial capitals of their respective countries. To compare the neighborhoods of the two cities and determine how similar or dissimilar they are by the venue of categories they have in the city.

## Data <a name="data"></a>

Based on definition of our problem, factors that will influence our decission are:
* Venue data of the Toronto
* Venue data of the New York
* Total number of different kinds of venue of the two cities
* Number of different kinds of venue in the neighborhood of the two cities


We decided to use postcod to define our neighborhoods.

Following data sources will be needed to extract/generate the required information:
* The latitude and longitude data of those areas will be obtained using **Google Maps API reverse geocoding**
* number of venue and their type and location in every neighborhood will be obtained using **Foursquare API**

### Neighborhood Candidates

Let's create latitude & longitude coordinates for centroids of all neighborhoods of the two cities. We will create a grid of cells covering our area of interest which is aprox. 0.5 killometers centered around every neighbourhoods.

Let's first find the latitude & longitude of the neighbourhoods, using specific, well known address and Google Maps geocoding API.

# 0. Installing the pre-requisted libs

In [1]:
!pip install beautifulsoup4 requests pandas geocoder

     |████████████████████████████████| 102kB 15.0MB/s ta 0:00:01


## 1. Prepared data for Toronto

### 1.1 Data scrape for suburb and postcode

In [2]:
from bs4 import BeautifulSoup
import requests
import pandas as pd
import geocoder
import time
import json

### 1.1 Scraping from wikipedia

In [3]:
def link_or_text(elem):
    link = elem.select_one('a')
    if link:
        return link.text.strip()
    return elem.text.strip()

In [4]:
URL = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
pc_data = []

html = requests.get(URL).content
soup = BeautifulSoup(html, 'html.parser')
tbody = soup.select_one('table.wikitable>tbody')

for tr in tbody.find_all('tr'):
    tds = tr.find_all('td')
    if len(tds) == 3:
        pc, borough, neighbourhood = [link_or_text(td) for td in tds]
        if borough == 'Not assigned':
            # ignore when borough is ''Not assigned
            continue
        item = dict(zip(['Postcode', 'Borough', 'Neighbourhood'],  [pc, borough, neighbourhood]))
        pc_data.append(item) 

In [5]:
pc_data[:3]

[{'Postcode': 'M3A', 'Borough': 'North York', 'Neighbourhood': 'Parkwoods'},
 {'Postcode': 'M4A',
  'Borough': 'North York',
  'Neighbourhood': 'Victoria Village'},
 {'Postcode': 'M5A',
  'Borough': 'Downtown Toronto',
  'Neighbourhood': 'Harbourfront'}]

In [6]:
pc_df = pd.DataFrame(pc_data)

In [7]:
pc_df.head()

Borough     Neighbourhood Postcode
0        North York         Parkwoods      M3A
1        North York  Victoria Village      M4A
2  Downtown Toronto      Harbourfront      M5A
3  Downtown Toronto       Regent Park      M5A
4        North York  Lawrence Heights      M6A

In [8]:
pc_df.tail()

Borough             Neighbourhood Postcode
206  Etobicoke  Kingsway Park South West      M8Z
207  Etobicoke                 Mimico NW      M8Z
208  Etobicoke        The Queensway West      M8Z
209  Etobicoke     Royal York South West      M8Z
210  Etobicoke            South of Bloor      M8Z

In [9]:
pc_df = pc_df[pc_df.Borough != 'Not assigned']

In [10]:
pc_df = pc_df[['Postcode','Borough','Neighbourhood']]
pc_df.head()

Postcode           Borough     Neighbourhood
0      M3A        North York         Parkwoods
1      M4A        North York  Victoria Village
2      M5A  Downtown Toronto      Harbourfront
3      M5A  Downtown Toronto       Regent Park
4      M6A        North York  Lawrence Heights

In [11]:
pc_df_2 = pc_df.groupby('Postcode')['Neighbourhood'].apply(', '.join).reset_index()
pc_df_2.head()

Postcode                           Neighbourhood
0      M1B                          Rouge, Malvern
1      M1C  Highland Creek, Rouge Hill, Port Union
2      M1E       Guildwood, Morningside, West Hill
3      M1G                                  Woburn
4      M1H                               Cedarbrae

In [12]:
pc_df_3 = pc_df[['Postcode','Borough']]


In [13]:
pc_df_2 = pd.merge(pc_df_3, pc_df_2, on='Postcode', how='inner') 
pc_df_2 = pc_df_2.drop_duplicates(subset=['Postcode', 'Borough','Neighbourhood'], keep='first')
pc_df_2.head()

Postcode           Borough                     Neighbourhood
0      M3A        North York                         Parkwoods
1      M4A        North York                  Victoria Village
2      M5A  Downtown Toronto         Harbourfront, Regent Park
4      M6A        North York  Lawrence Heights, Lawrence Manor
6      M7A      Queen's Park                      Not assigned

In [14]:
for index, row in pc_df_2.iterrows():
    if row['Neighbourhood']=='Not assigned': 
        row['Neighbourhood']=row['Borough']


In [15]:
pc_df_2.head()

Postcode           Borough                     Neighbourhood
0      M3A        North York                         Parkwoods
1      M4A        North York                  Victoria Village
2      M5A  Downtown Toronto         Harbourfront, Regent Park
4      M6A        North York  Lawrence Heights, Lawrence Manor
6      M7A      Queen's Park                      Queen's Park

In [16]:
pc_df.head()

Postcode           Borough     Neighbourhood
0      M3A        North York         Parkwoods
1      M4A        North York  Victoria Village
2      M5A  Downtown Toronto      Harbourfront
3      M5A  Downtown Toronto       Regent Park
4      M6A        North York  Lawrence Heights

### 1.2 Retriving location coordinates

In [17]:
GCP_API_KEY = 'THIS_IS_A_SECRET'

In [18]:
#@hidden_cell
GCP_API_KEY = 'AIzaSyC4AdT2WqA17f8ZEO5K_TbOWRV3wW4A8m4'

In [19]:
latlong_data = []

qry_result = []
for (idx, pc) in pc_df_2['Postcode'].items():
    # pc = item.get('Postcode')
    qry = f'{pc}, Toronto, Ontario'
    for i in range(3):
        r = geocoder.google(qry, key=GCP_API_KEY)
        if r.latlng:
            #print(f'{qry} {r.latlng}')
            qry_result.append([pc, r.latlng[0], r.latlng[1]])
            break
        else:
            time.sleep(1.5)
    

In [20]:
## `latlongs` is just a list of dictionary with keys Latitude and Longtitude 
## and we will merge it back to the pc_data later
latlongs = [dict([('Postcode', r[0]), ('Latitude', r[1]), ('Longtitude', r[2])])  for r in  qry_result]

In [21]:
len(latlongs), len(pc_data)

(103, 211)

In [22]:
latlongs[:5]

[{'Postcode': 'M3A', 'Latitude': 43.7532586, 'Longtitude': -79.3296565},
 {'Postcode': 'M4A', 'Latitude': 43.72588229999999, 'Longtitude': -79.3155716},
 {'Postcode': 'M5A', 'Latitude': 43.6542599, 'Longtitude': -79.36063589999999},
 {'Postcode': 'M6A', 'Latitude': 43.718518, 'Longtitude': -79.4647633},
 {'Postcode': 'M7A', 'Latitude': 43.6623015, 'Longtitude': -79.3894938}]

In [23]:
## Merging the pc_data and latlongs
_ = [d.update(d_latlng) for (d, d_latlng) in zip(pc_data, latlongs)]

In [24]:
df_latlongs = pd.DataFrame(latlongs)

In [25]:
df_latlongs

Latitude  Longtitude Postcode
0    43.753259  -79.329656      M3A
1    43.725882  -79.315572      M4A
2    43.654260  -79.360636      M5A
3    43.718518  -79.464763      M6A
4    43.662301  -79.389494      M7A
5    43.667856  -79.532242      M9A
6    43.806686  -79.194353      M1B
7    43.745906  -79.352188      M3B
8    43.706397  -79.309937      M4B
9    43.657162  -79.378937      M5B
10   43.709577  -79.445073      M6B
11   43.650943  -79.554724      M9B
12   43.784535  -79.160497      M1C
13   43.725900  -79.340923      M3C
14   43.695344  -79.318389      M4C
15   43.651494  -79.375418      M5C
16   43.693781  -79.428191      M6C
17   43.643515  -79.577201      M9C
18   43.763573  -79.188711      M1E
19   43.676357  -79.293031      M4E
20   43.644771  -79.373306      M5E
21   43.689026  -79.453512      M6E
22   43.770992  -79.216917      M1G
23   43.709060  -79.363452      M4G
24   43.657952  -79.387383      M5G
25   43.669542  -79.422564      M6G
26   43.773136  -79.239476      M1H
27   43.803762  -79.363452      M2H
28   43.754328  -79.442259      M3H
29   43.705369  -79.349372      M4H
..         ...         ...      ...
73   43.715383  -79.405678      M4R
74   43.672710  -79.405678      M5R
75   43.648960  -79.456325      M6R
76   43.636966  -79.615819      M7R
77   43.688905  -79.554724      M9R
78   43.794200  -79.262029      M1S
79   43.704324  -79.388790      M4S
80   43.662696  -79.400049      M5S
81   43.651571  -79.484450      M6S
82   43.781638  -79.304302      M1T
83   43.689574  -79.383160      M4T
84   43.653206  -79.400049      M5T
85   43.815252  -79.284577      M1V
86   43.686412  -79.400049      M4V
87   43.628947  -79.394420      M5V
88   43.605647  -79.501321      M8V
89   43.739416  -79.588437      M9V
90   43.799525  -79.318389      M1W
91   43.679563  -79.377529      M4W
92   43.646435  -79.374846      M5W
93   43.602414  -79.543484      M8W
94   43.706748  -79.594054      M9W
95   43.836125  -79.205636      M1X
96   43.667967  -79.367675      M4X
97   43.648429  -79.382280      M5X
98   43.653654  -79.506944      M8X
99   43.665860  -79.383160      M4Y
100  43.662744  -79.321558      M7Y
101  43.636258  -79.498509      M8Y
102  43.628841  -79.520999      M8Z

[103 rows x 3 columns]

In [73]:
df = pd.merge(pc_df_2, df_latlongs, on='Postcode', how='left')
df.head()

Postcode           Borough                     Neighbourhood   Latitude  \
0      M3A        North York                         Parkwoods  43.753259   
1      M4A        North York                  Victoria Village  43.725882   
2      M5A  Downtown Toronto         Harbourfront, Regent Park  43.654260   
3      M6A        North York  Lawrence Heights, Lawrence Manor  43.718518   
4      M7A      Queen's Park                      Queen's Park  43.662301   

   Longtitude  
0  -79.329656  
1  -79.315572  
2  -79.360636  
3  -79.464763  
4  -79.389494

In [74]:
df2= pd.merge(pc_df, df_latlongs, on='Postcode', how='left')
df2.head()

Postcode           Borough     Neighbourhood   Latitude  Longtitude
0      M3A        North York         Parkwoods  43.753259  -79.329656
1      M4A        North York  Victoria Village  43.725882  -79.315572
2      M5A  Downtown Toronto      Harbourfront  43.654260  -79.360636
3      M5A  Downtown Toronto       Regent Park  43.654260  -79.360636
4      M6A        North York  Lawrence Heights  43.718518  -79.464763

### 1.3 Get venue info

### Define Foursquare Credentials and Version


In [28]:
CLIENT_ID = 'FWQGJHQKEPGJI5QSSD01TZYJB12ZR20ZGHJY4FR130RRHTJI' # your Foursquare ID
CLIENT_SECRET = 'FX3YOPGXYNF5JBSD445TBPOONT4WZ54I5GTQSIKMO4RCNN1A' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: FWQGJHQKEPGJI5QSSD01TZYJB12ZR20ZGHJY4FR130RRHTJI
CLIENT_SECRET:FX3YOPGXYNF5JBSD445TBPOONT4WZ54I5GTQSIKMO4RCNN1A


In [29]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [30]:
LIMIT=100

In [75]:
toronto_venues = getNearbyVenues(names=df['Neighbourhood'],
                                   latitudes=df['Latitude'],
                                   longitudes=df['Longtitude']
                                  )

Parkwoods
Victoria Village
Harbourfront, Regent Park
Lawrence Heights, Lawrence Manor
Queen's Park
Islington Avenue
Rouge, Malvern
Don Mills North
Woodbine Gardens, Parkview Hill
Ryerson, Garden District
Glencairn
Cloverdale, Islington, Martin Grove, Princess Gardens, West Deane Park
Highland Creek, Rouge Hill, Port Union
Flemingdon Park, Don Mills South
Woodbine Heights
St. James Town
Humewood-Cedarvale
Bloordale Gardens, Eringate, Markland Wood, Old Burnhamthorpe
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Downsview North, Wilson Heights
Thorncliffe Park
Adelaide, King, Richmond
Dovercourt Village, Dufferin
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto
Harbourfront East, Toronto Islands, Union Station
Little Portugal, Trinity
East Birchmount Park, Ionview, Kennedy Park
Bayview Village
CFB Toronto, Downsview East
The D

In [76]:
toronto_venues_save = toronto_venues

In [77]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                                              100   
Agincourt                                                               5   
Agincourt North, L'Amoreaux East, Milliken, Ste...                      3   
Albion Gardens, Beaumond Heights, Humbergate, J...                      9   
Alderwood, Long Branch                                                 10   
Bathurst Manor, Downsview North, Wilson Heights                        18   
Bayview Village                                                         4   
Bedford Park, Lawrence Manor East                                      25   
Berczy Park                                                            57   
Birch Cliff, Cliffside West                                             4   
Bloordale Gardens, Eringate, Markland Wood, Old...                      9   
Brockton, Exhibition Place, Parkdale Village                           22   
Business Reply Mail Processing Centre 969 Eastern                      17   
CFB Toronto, Downsview East                                             2   
CN Tower, Bathurst Quay, Island airport, Harbou...                     15   
Cabbagetown, St. James Town                                            47   
Caledonia-Fairbanks                                                     5   
Canada Post Gateway Processing Centre                                  10   
Cedarbrae                                                               7   
Central Bay Street                                                     85   
Chinatown, Grange Park, Kensington Market                             100   
Christie                                                               15   
Church and Wellesley                                                   91   
Clairlea, Golden Mile, Oakridge                                        10   
Clarks Corners, Sullivan, Tam O'Shanter                                13   
Cliffcrest, Cliffside, Scarborough Village West                         3   
Cloverdale, Islington, Martin Grove, Princess G...                      1   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             33   
Davisville North                                                        8   
...                                                                   ...   
Northwood Park, York University                                         7   
Parkdale, Roncesvalles                                                 15   
Parkwoods                                                               3   
Queen's Park                                                           42   
Rosedale                                                                5   
Roselawn                                                                2   
Rouge, Malvern                                                          1   
Runnymede, Swansea                                                     37   
Ryerson, Garden District                                              100   
Scarborough Village                                                     1   
Silver Hills, York Mills                                                1   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        98   
Studio District                                                        38   
The Annex, North Midtown, Yorkville                                    23   
The Beaches                                                             4   
The Beaches West, India Bazaar                                         22   
The Danforth West, Riverdale                                           44   
The Junction North, Runnymede                                           3   
The Kingsway, Montgomery Road, Old Mill North   

In [78]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 278 uniques categories.


In [79]:
toronto_venues

Neighborhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                             Parkwoods   
3                                      Victoria Village   
4                                      Victoria Village   
5                                      Victoria Village   
6                                      Victoria Village   
7                             Harbourfront, Regent Park   
8                             Harbourfront, Regent Park   
9                             Harbourfront, Regent Park   
10                            Harbourfront, Regent Park   
11                            Harbourfront, Regent Park   
12                            Harbourfront, Regent Park   
13                            Harbourfront, Regent Park   
14                            Harbourfront, Regent Park   
15                            Harbourfront, Regent Park   
16                            Harbourfront, Regent Park   
17                            Harbourfront, Regent Park   
18                            Harbourfront, Regent Park   
19                            Harbourfront, Regent Park   
20                            Harbourfront, Regent Park   
21                            Harbourfront, Regent Park   
22                            Harbourfront, Regent Park   
23                            Harbourfront, Regent Park   
24                            Harbourfront, Regent Park   
25                            Harbourfront, Regent Park   
26                            Harbourfront, Regent Park   
27                            Harbourfront, Regent Park   
28                            Harbourfront, Regent Park   
29                            Harbourfront, Regent Park   
...                                                 ...   
2234  Business Reply Mail Processing Centre 969 Eastern   
2235  Business Reply Mail Processing Centre 969 Eastern   
2236  Business Reply Mail Processing Centre 969 Eastern   
2237  Business Reply Mail Processing Centre 969 Eastern   
2238  Business Reply Mail Processing Centre 969 Eastern   
2239  Business Reply Mail Processing Centre 969 Eastern   
2240  Business Reply Mail Processing Centre 969 Eastern   
2241  Business Reply Mail Processing Centre 969 Eastern   
2242  Business Reply Mail Processing Centre 969 Eastern   
2243  Business Reply Mail Processing Centre 969 Eastern   
2244  Business Reply Mail Processing Centre 969 Eastern   
2245  Business Reply Mail Processing Centre 969 Eastern   
2246  Business Reply Mail Processing Centre 969 Eastern   
2247  Humber Bay, King's Mill Park, Kingsway Park So...   
2248  Humber Bay, King's Mill Park, Kingsway Park So...   
2249  Humber Bay, King's Mill Park, Kingsway Park So...   
2250  Kingsway Park South West, Mimico NW, The Queen...   
2251  Kingsway Park South West, Mimico NW, The Queen...   
2252  Kingsway Park South West, Mimico NW, The Queen...   
2253  Kingsway Park South West, Mimico NW, The Queen...   
2254  Kingsway Park South West, Mimico NW, The Queen...   
2255  Kingsway Park South West, Mimico NW, The Queen...   
2256  Kingsway Park South West, Mimico NW, The Queen...   
2257  Kingsway Park South West, Mimico NW, The Queen...   
2258  Kingsway Park South West, Mimico NW, The Queen...   
2259  Kingsway Park South West, Mimico NW, The Queen...   
2260  Kingsway Park South West, Mimico NW, The Queen...   
2261  Kingsway Park South West, Mimico NW, The Queen...   
2262  Kingsway Park South West, Mimico NW, The Queen...   
2263  Kingsway Park South West, Mimico NW, The Queen...   

      Neighborhood Latitude  Neighborhood Longitude  \
0                 43.753259              -79.329656   
1                 43.753259              -79.329656   
2                 43.753259              -79.329656   
3                 43.725882              -79.315572   
4                 43.725882              -79.315572   
5                 43.725882              -79.315572   

In [80]:
toronto_venues.to_csv('C:\\Users\\OSS\\Downloads\\toronto.csv', sep='\t', encoding='utf-8')

In [81]:
toronto_venues_d = toronto_venues.groupby(['Neighborhood','Venue Category']).count()

In [83]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues_save[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues_save['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Accessories Store  Afghan Restaurant  Airport  \
0            0                  0                  0        0   
1            0                  0                  0        0   
2            0                  0                  0        0   
3            0                  0                  0        0   
4            0                  0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Trail  Train Station  \
0                 0                    0  ...      0              0   
1                 0                    0  ...      0              0   
2                 0                    0  ...      0              0   
3                 0                    0  ...      0              0   
4                 0                    0  ...      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                              0                 0            0   
1                              0                 0            0   
2                              0                 0            0   
3                              0                 0            0   
4                              0                 0            0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  
0              0  
1              0  
2              0  
3              0  
4              0  

[5 rows x 278 columns]

In [85]:
toronto_grouped = toronto_onehot.groupby(['Neighborhood']).mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                            Adelaide, King, Richmond     0.000000   
1                                           Agincourt     0.000000   
2   Agincourt North, L'Amoreaux East, Milliken, St...     0.000000   
3   Albion Gardens, Beaumond Heights, Humbergate, ...     0.000000   
4                              Alderwood, Long Branch     0.000000   
5     Bathurst Manor, Downsview North, Wilson Heights     0.000000   
6                                     Bayview Village     0.000000   
7                   Bedford Park, Lawrence Manor East     0.000000   
8                                         Berczy Park     0.000000   
9                         Birch Cliff, Cliffside West     0.000000   
10  Bloordale Gardens, Eringate, Markland Wood, Ol...     0.000000   
11       Brockton, Exhibition Place, Parkdale Village     0.000000   
12  Business Reply Mail Processing Centre 969 Eastern     0.058824   
13                        CFB Toronto, Downsview East     0.000000   
14  CN Tower, Bathurst Quay, Island airport, Harbo...     0.000000   
15                        Cabbagetown, St. James Town     0.000000   
16                                Caledonia-Fairbanks     0.000000   
17              Canada Post Gateway Processing Centre     0.000000   
18                                          Cedarbrae     0.000000   
19                                 Central Bay Street     0.011765   
20          Chinatown, Grange Park, Kensington Market     0.000000   
21                                           Christie     0.000000   
22                               Church and Wellesley     0.010989   
23                    Clairlea, Golden Mile, Oakridge     0.000000   
24            Clarks Corners, Sullivan, Tam O'Shanter     0.000000   
25    Cliffcrest, Cliffside, Scarborough Village West     0.000000   
26  Cloverdale, Islington, Martin Grove, Princess ...     0.000000   
27                     Commerce Court, Victoria Hotel     0.000000   
28                                         Davisville     0.000000   
29                                   Davisville North     0.000000   
..                                                ...          ...   
70                    Northwood Park, York University     0.000000   
71                             Parkdale, Roncesvalles     0.000000   
72                                          Parkwoods     0.000000   
73                                       Queen's Park     0.023810   
74                                           Rosedale     0.000000   
75                                           Roselawn     0.000000   
76                                     Rouge, Malvern     0.000000   
77                                 Runnymede, Swansea     0.000000   
78                           Ryerson, Garden District     0.000000   
79                                Scarborough Village     0.000000   
80                           Silver Hills, York Mills     0.000000   
81                                     St. James Town     0.000000   
82                    Stn A PO Boxes 25 The Esplanade     0.000000   
83                                    Studio District     0.026316   
84                The Annex, North Midtown, Yorkville     0.000000   
85                                        The Beaches     0.000000   
86                     The Beaches West, India Bazaar     0.000000   
87                       The Danforth West, Riverdale     0.022727   
88                      The Junction North, Runnymede     0.000000   
89      The Kingsway, Montgomery Road, Old Mill North     0.000000   
90                                   Thorncliffe Park     0.052632   
91                                   Victoria Village     0.000000   
92                                          Westmount     0.000000   
93                                             Weston     0.000000   
94                                   Willowdale South     0.000000   
95                                    Willowdale We

### 2. Prepare data for New York

In [41]:
!wget -q -O 'newyork_data.json' https://cocl.us/new_york_dataset
print('Data downloaded!')

Data downloaded!


In [42]:
with open('newyork_data.json') as json_data:
    newyork_data = json.load(json_data)

In [43]:
neighborhoods_data = newyork_data['features']

In [44]:
neighborhoods_data[0]

{'type': 'Feature',
 'id': 'nyu_2451_34572.1',
 'geometry': {'type': 'Point',
  'coordinates': [-73.84720052054902, 40.89470517661]},
 'geometry_name': 'geom',
 'properties': {'name': 'Wakefield',
  'stacked': 1,
  'annoline1': 'Wakefield',
  'annoline2': None,
  'annoline3': None,
  'annoangle': 0.0,
  'borough': 'Bronx',
  'bbox': [-73.84720052054902,
   40.89470517661,
   -73.84720052054902,
   40.89470517661]}}

In [45]:
# define the dataframe columns
column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude'] 

# instantiate the dataframe
neighborhoods = pd.DataFrame(columns=column_names)

In [46]:
for data in neighborhoods_data:
    borough = neighborhood_name = data['properties']['borough'] 
    neighborhood_name = data['properties']['name']
        
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[1]
    neighborhood_lon = neighborhood_latlon[0]
    
    neighborhoods = neighborhoods.append({'Borough': borough,
                                          'Neighborhood': neighborhood_name,
                                          'Latitude': neighborhood_lat,
                                          'Longitude': neighborhood_lon}, ignore_index=True)

In [47]:
manhattan_data = neighborhoods[neighborhoods['Borough'] == 'Manhattan'].reset_index(drop=True)
manhattan_data.head()

Borough        Neighborhood   Latitude  Longitude
0  Manhattan         Marble Hill  40.876551 -73.910660
1  Manhattan           Chinatown  40.715618 -73.994279
2  Manhattan  Washington Heights  40.851903 -73.936900
3  Manhattan              Inwood  40.867684 -73.921210
4  Manhattan    Hamilton Heights  40.823604 -73.949688

In [48]:
neighborhoods.head()

Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939
2   Bronx  Eastchester  40.887556 -73.827806
3   Bronx    Fieldston  40.895437 -73.905643
4   Bronx    Riverdale  40.890834 -73.912585

In [49]:
manhattan_venues = getNearbyVenues(names=manhattan_data['Neighborhood'],
                                   latitudes=manhattan_data['Latitude'],
                                   longitudes=manhattan_data['Longitude']
                                  )

Marble Hill
Chinatown
Washington Heights
Inwood
Hamilton Heights
Manhattanville
Central Harlem
East Harlem
Upper East Side
Yorkville
Lenox Hill
Roosevelt Island
Upper West Side
Lincoln Square
Clinton
Midtown
Murray Hill
Chelsea
Greenwich Village
East Village
Lower East Side
Tribeca
Little Italy
Soho
West Village
Manhattan Valley
Morningside Heights
Gramercy
Battery Park City
Financial District
Carnegie Hill
Noho
Civic Center
Midtown South
Sutton Place
Turtle Bay
Tudor City
Stuyvesant Town
Flatiron
Hudson Yards


In [50]:
# one hot encoding
manhattan_onehot = pd.get_dummies(manhattan_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
manhattan_onehot['Neighborhood'] = manhattan_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [manhattan_onehot.columns[-1]] + list(manhattan_onehot.columns[:-1])
manhattan_onehot = manhattan_onehot[fixed_columns]

manhattan_onehot.head()

Neighborhood  Accessories Store  Adult Boutique  Afghan Restaurant  \
0  Marble Hill                  0               0                  0   
1  Marble Hill                  0               0                  0   
2  Marble Hill                  0               0                  0   
3  Marble Hill                  0               0                  0   
4  Marble Hill                  0               0                  0   

   African Restaurant  American Restaurant  Animal Shelter  Antique Shop  \
0                   0                    0               0             0   
1                   0                    0               0             0   
2                   0                    0               0             0   
3                   0                    0               0             0   
4                   0                    0               0             0   

   Arcade  Arepa Restaurant  ...  Vietnamese Restaurant  Volleyball Court  \
0       0                 0  ...                      0                 0   
1       0                 0  ...                      0                 0   
2       0                 0  ...                      0                 0   
3       0                 0  ...                      0                 0   
4       0                 0  ...                      0                 0   

   Waterfront  Weight Loss Center  Whisky Bar  Wine Bar  Wine Shop  \
0           0                   0           0         0          0   
1           0                   0           0         0          0   
2           0                   0           0         0          0   
3           0                   0           0         0          0   
4           0                   0           0         0          0   

   Wings Joint  Women's Store  Yoga Studio  
0            0              0            0  
1            0              0            1  
2            0              0            0  
3            0              0            0  
4            0              0            0  

[5 rows x 339 columns]

In [87]:
manhattan_grouped = manhattan_onehot.groupby('Neighborhood').mean().reset_index()
manhattan_grouped

Neighborhood  Accessories Store  Adult Boutique  Afghan Restaurant  \
0     Battery Park City           0.000000            0.00               0.00   
1         Carnegie Hill           0.000000            0.00               0.00   
2        Central Harlem           0.000000            0.00               0.00   
3               Chelsea           0.000000            0.00               0.00   
4             Chinatown           0.000000            0.00               0.00   
5          Civic Center           0.000000            0.00               0.00   
6               Clinton           0.000000            0.00               0.00   
7           East Harlem           0.000000            0.00               0.00   
8          East Village           0.000000            0.00               0.00   
9    Financial District           0.010000            0.00               0.00   
10             Flatiron           0.000000            0.00               0.00   
11             Gramercy           0.000000            0.00               0.00   
12    Greenwich Village           0.000000            0.00               0.00   
13     Hamilton Heights           0.000000            0.00               0.00   
14         Hudson Yards           0.000000            0.00               0.00   
15               Inwood           0.000000            0.00               0.00   
16           Lenox Hill           0.000000            0.00               0.01   
17       Lincoln Square           0.000000            0.00               0.00   
18         Little Italy           0.000000            0.00               0.00   
19      Lower East Side           0.000000            0.00               0.00   
20     Manhattan Valley           0.000000            0.00               0.00   
21       Manhattanville           0.000000            0.00               0.00   
22          Marble Hill           0.000000            0.00               0.00   
23              Midtown           0.000000            0.00               0.00   
24        Midtown South           0.000000            0.00               0.00   
25  Morningside Heights           0.000000            0.00               0.00   
26          Murray Hill           0.000000            0.00               0.00   
27                 Noho           0.000000            0.01               0.00   
28     Roosevelt Island           0.000000            0.00               0.00   
29                 Soho           0.010000            0.00               0.00   
30      Stuyvesant Town           0.000000            0.00               0.00   
31         Sutton Place           0.000000            0.01               0.00   
32              Tribeca           0.000000            0.00               0.00   
33           Tudor City           0.000000            0.00               0.00   
34           Turtle Bay           0.000000            0.00               0.00   
35      Upper East Side           0.000000            0.00               0.00   
36      Upper West Side           0.010000            0.00               0.00   
37   Washington Heights           0.011628            0.00               0.00   
38         West Village           0.010000            0.00               0.00   
39            Yorkville           0.000000            0.00               0.00   

    African Restaurant  American Restaurant  Animal Shelter  Antique Shop  \
0             0.000000             0.010000            0.00          0.00   
1             0.000000             0.010000            0.00          0.00   
2             0.065217             0.043478            0.00          0.00   
3             0.000000             0.030000            0.00          0.00   
4             0.000000             0.040000            0.00          0.00   
5             0.000000             0.030000            0.00          0.01   
6             0.000000             0.040000            0.00          0.00   
7             0.000000             0.000000            0.00          0.00

In [52]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods['Borough'].unique()),
        neighborhoods.shape[0]
    )
)

The dataframe has 5 boroughs and 306 neighborhoods.


## Methodology <a name="methodology"></a>

In this project we will compare Toronto and New York by their neigbourhoods using different catolgories vene data.

For each suburbs, compare the number of different venues and use K-mean Clustering to group the neigbourhoods to see if there is silificant differences between to cities. 


## 3. k-means Clustering

#### Define a function that assigns each datapoint to a cluster

In [54]:
import numpy as np

In [55]:
colors_map = np.array(['b', 'r'])
def assign_members(x1, x2, centers):
    compare_to_first_center = np.sqrt(np.square(np.array(x1) - centers[0][0]) + np.square(np.array(x2) - centers[0][1]))
    compare_to_second_center = np.sqrt(np.square(np.array(x1) - centers[1][0]) + np.square(np.array(x2) - centers[1][1]))
    class_of_points = compare_to_first_center > compare_to_second_center
    colors = colors_map[class_of_points + 1 - 1]
    return colors, class_of_points

print('assign_members function defined!')

assign_members function defined!


#### Define a function that updates the centroid of each cluster

In [56]:
# update means
def update_centers(x1, x2, class_of_points):
    center1 = [np.mean(np.array(x1)[~class_of_points]), np.mean(np.array(x2)[~class_of_points])]
    center2 = [np.mean(np.array(x1)[class_of_points]), np.mean(np.array(x2)[class_of_points])]
    return [center1, center2]

print('assign_members function defined!')

assign_members function defined!


#### Define a function that plots the data points along with the cluster centroids

In [57]:
def plot_points(centroids=None, colors='g', figure_title=None):
    # plot the figure
    fig = plt.figure(figsize=(15, 10))  # create a figure object
    ax = fig.add_subplot(1, 1, 1)
    
    centroid_colors = ['bx', 'rx']
    if centroids:
        for (i, centroid) in enumerate(centroids):
            ax.plot(centroid[0], centroid[1], centroid_colors[i], markeredgewidth=5, markersize=20)
    plt.scatter(x1, x2, s=500, c=colors)
    
    # define the ticks
    xticks = np.linspace(-6, 8, 15, endpoint=True)
    yticks = np.linspace(-6, 6, 13, endpoint=True)

    # fix the horizontal axis
    ax.set_xticks(xticks)
    ax.set_yticks(yticks)

    # add tick labels
    xlabels = xticks
    ax.set_xticklabels(xlabels)
    ylabels = yticks
    ax.set_yticklabels(ylabels)

    # style the ticks
    ax.xaxis.set_ticks_position('bottom')
    ax.yaxis.set_ticks_position('left')
    ax.tick_params('both', length=2, width=1, which='major', labelsize=15)
    
    # add labels to axes
    ax.set_xlabel('x1', fontsize=20)
    ax.set_ylabel('x2', fontsize=20)
    
    # add title to figure
    ax.set_title(figure_title, fontsize=24)

    plt.show()

print('plot_points function defined!')

plot_points function defined!


In [89]:
toronto_grouped_new = toronto_grouped

In [90]:
toronto_grouped_new['Neighborhood']= toronto_grouped_new['Neighborhood']+ "_t"

In [91]:
toronto_grouped_new

Neighborhood  Yoga Studio  \
0                          Adelaide, King, Richmond_t     0.000000   
1                                         Agincourt_t     0.000000   
2   Agincourt North, L'Amoreaux East, Milliken, St...     0.000000   
3   Albion Gardens, Beaumond Heights, Humbergate, ...     0.000000   
4                            Alderwood, Long Branch_t     0.000000   
5   Bathurst Manor, Downsview North, Wilson Heights_t     0.000000   
6                                   Bayview Village_t     0.000000   
7                 Bedford Park, Lawrence Manor East_t     0.000000   
8                                       Berczy Park_t     0.000000   
9                       Birch Cliff, Cliffside West_t     0.000000   
10  Bloordale Gardens, Eringate, Markland Wood, Ol...     0.000000   
11     Brockton, Exhibition Place, Parkdale Village_t     0.000000   
12  Business Reply Mail Processing Centre 969 East...     0.058824   
13                      CFB Toronto, Downsview East_t     0.000000   
14  CN Tower, Bathurst Quay, Island airport, Harbo...     0.000000   
15                      Cabbagetown, St. James Town_t     0.000000   
16                              Caledonia-Fairbanks_t     0.000000   
17            Canada Post Gateway Processing Centre_t     0.000000   
18                                        Cedarbrae_t     0.000000   
19                               Central Bay Street_t     0.011765   
20        Chinatown, Grange Park, Kensington Market_t     0.000000   
21                                         Christie_t     0.000000   
22                             Church and Wellesley_t     0.010989   
23                  Clairlea, Golden Mile, Oakridge_t     0.000000   
24          Clarks Corners, Sullivan, Tam O'Shanter_t     0.000000   
25  Cliffcrest, Cliffside, Scarborough Village West_t     0.000000   
26  Cloverdale, Islington, Martin Grove, Princess ...     0.000000   
27                   Commerce Court, Victoria Hotel_t     0.000000   
28                                       Davisville_t     0.000000   
29                                 Davisville North_t     0.000000   
..                                                ...          ...   
70                  Northwood Park, York University_t     0.000000   
71                           Parkdale, Roncesvalles_t     0.000000   
72                                        Parkwoods_t     0.000000   
73                                     Queen's Park_t     0.023810   
74                                         Rosedale_t     0.000000   
75                                         Roselawn_t     0.000000   
76                                   Rouge, Malvern_t     0.000000   
77                               Runnymede, Swansea_t     0.000000   
78                         Ryerson, Garden District_t     0.000000   
79                              Scarborough Village_t     0.000000   
80                         Silver Hills, York Mills_t     0.000000   
81                                   St. James Town_t     0.000000   
82                  Stn A PO Boxes 25 The Esplanade_t     0.000000   
83                                  Studio District_t     0.026316   
84              The Annex, North Midtown, Yorkville_t     0.000000   
85                                      The Beaches_t     0.000000   
86                   The Beaches West, India Bazaar_t     0.000000   
87                     The Danforth West, Riverdale_t     0.022727   
88                    The Junction North, Runnymede_t     0.000000   
89    The Kingsway, Montgomery Road, Old Mill North_t     0.000000   
90                                 Thorncliffe Park_t     0.052632   
91                                 Victoria Village_t     0.000000   
92                                        Westmount_t     0.000000   
93                                           Weston_t     0.000000   
94                                 Willowdale South_t     0.000000   
95                                  Willowdale West

In [92]:
manhattan_grouped_new = manhattan_grouped

In [93]:
manhattan_grouped_new['Neighborhood']=manhattan_grouped_new['Neighborhood']+'_n'

In [94]:
manhattan_grouped_new

Neighborhood  Accessories Store  Adult Boutique  \
0     Battery Park City_n           0.000000            0.00   
1         Carnegie Hill_n           0.000000            0.00   
2        Central Harlem_n           0.000000            0.00   
3               Chelsea_n           0.000000            0.00   
4             Chinatown_n           0.000000            0.00   
5          Civic Center_n           0.000000            0.00   
6               Clinton_n           0.000000            0.00   
7           East Harlem_n           0.000000            0.00   
8          East Village_n           0.000000            0.00   
9    Financial District_n           0.010000            0.00   
10             Flatiron_n           0.000000            0.00   
11             Gramercy_n           0.000000            0.00   
12    Greenwich Village_n           0.000000            0.00   
13     Hamilton Heights_n           0.000000            0.00   
14         Hudson Yards_n           0.000000            0.00   
15               Inwood_n           0.000000            0.00   
16           Lenox Hill_n           0.000000            0.00   
17       Lincoln Square_n           0.000000            0.00   
18         Little Italy_n           0.000000            0.00   
19      Lower East Side_n           0.000000            0.00   
20     Manhattan Valley_n           0.000000            0.00   
21       Manhattanville_n           0.000000            0.00   
22          Marble Hill_n           0.000000            0.00   
23              Midtown_n           0.000000            0.00   
24        Midtown South_n           0.000000            0.00   
25  Morningside Heights_n           0.000000            0.00   
26          Murray Hill_n           0.000000            0.00   
27                 Noho_n           0.000000            0.01   
28     Roosevelt Island_n           0.000000            0.00   
29                 Soho_n           0.010000            0.00   
30      Stuyvesant Town_n           0.000000            0.00   
31         Sutton Place_n           0.000000            0.01   
32              Tribeca_n           0.000000            0.00   
33           Tudor City_n           0.000000            0.00   
34           Turtle Bay_n           0.000000            0.00   
35      Upper East Side_n           0.000000            0.00   
36      Upper West Side_n           0.010000            0.00   
37   Washington Heights_n           0.011628            0.00   
38         West Village_n           0.010000            0.00   
39            Yorkville_n           0.000000            0.00   

    Afghan Restaurant  African Restaurant  American Restaurant  \
0                0.00            0.000000             0.010000   
1                0.00            0.000000             0.010000   
2                0.00            0.065217             0.043478   
3                0.00            0.000000             0.030000   
4                0.00            0.000000             0.040000   
5                0.00            0.000000             0.030000   
6                0.00            0.000000             0.040000   
7                0.00            0.000000             0.000000   
8                0.00            0.000000             0.020000   
9                0.00            0.000000             0.030000   
10               0.00            0.000000             0.040000   
11               0.00            0.000000             0.040000   
12               0.00            0.000000             0.020000   
13               0.00            0.000000             0.000000   
14               0.00            0.000000             0.064103   
15               0.00            0.000000             0.035088   
16               0.01            0.000000             0.000000   
17               0.00            0.000000             0.020000   
18               0.00            0.000000             0.010000   
19               0.00            0.000000             0.000000   
20

In [110]:
toronto_grouped_new.shape

(100, 278)

In [111]:
manhattan_grouped_new.shape

(40, 339)

## Analysis <a name="analysis"></a>

## Venue in Toronto not in Manhattan

In [100]:
toronto_grouped_new.columns.difference(manhattan_grouped_new.columns)

Index(['Airport', 'Airport Food Court', 'Airport Gate', 'Airport Lounge',
       'Airport Service', 'Airport Terminal', 'Aquarium', 'Auto Garage',
       'Baseball Stadium', 'Basketball Stadium', 'Beach', 'Belgian Restaurant',
       'Brewery', 'Check Cashing Service', 'Church', 'College Arts Building',
       'College Auditorium', 'College Gym', 'College Rec Center',
       'College Stadium', 'Colombian Restaurant', 'Comfort Food Restaurant',
       'Comic Shop', 'Construction & Landscaping', 'Curling Ice',
       'Doner Restaurant', 'Festival', 'Fish & Chips Shop', 'Food',
       'Fraternity House', 'Fruit & Vegetable Store', 'General Travel',
       'Gluten-free Restaurant', 'Hakka Restaurant', 'Hockey Arena',
       'Home Service', 'Hospital', 'Housing Development',
       'Indonesian Restaurant', 'Lake', 'Lawyer', 'Light Rail Station',
       'Luggage Store', 'Mac & Cheese Joint', 'Motel', 'Other Great Outdoors',
       'Plane', 'Polish Restaurant', 'Portuguese Restaurant', 'Pouti

## Venue in Manhattan not in Toronto

In [99]:
manhattan_grouped_new.columns.difference(toronto_grouped_new.columns)

Index(['Adult Boutique', 'African Restaurant', 'Animal Shelter', 'Arcade',
       'Arepa Restaurant', 'Argentinian Restaurant', 'Auditorium',
       'Australian Restaurant', 'Austrian Restaurant', 'Beer Garden',
       ...
       'Turkish Restaurant', 'Udon Restaurant', 'Used Bookstore',
       'Venezuelan Restaurant', 'Veterinarian', 'Volleyball Court',
       'Waterfront', 'Weight Loss Center', 'Whisky Bar', 'Wine Shop'],
      dtype='object', length=122)

## Restaurant tpye

In [108]:
kind_t = toronto_grouped_new.columns.str.contains(r'Restaurant')
unique, counts = np.unique(kind_t, return_counts=True)
dict(zip(unique, counts))

{False: 225, True: 53}

In [112]:
kind_m = manhattan_grouped_new.columns.str.contains(r'Restaurant')
unique_m, counts_m = np.unique(kind_m, return_counts=True)
dict(zip(unique_m, counts_m))

{False: 263, True: 76}

In [160]:
total_group = pd.concat([toronto_grouped_new,manhattan_grouped_new])

/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:1: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  if __name__ == '__main__':


In [161]:
total_group

Accessories Store  Adult Boutique  Afghan Restaurant  African Restaurant  \
0            0.000000             NaN           0.000000                 NaN   
1            0.000000             NaN           0.000000                 NaN   
2            0.000000             NaN           0.000000                 NaN   
3            0.000000             NaN           0.000000                 NaN   
4            0.000000             NaN           0.000000                 NaN   
5            0.000000             NaN           0.000000                 NaN   
6            0.000000             NaN           0.000000                 NaN   
7            0.000000             NaN           0.000000                 NaN   
8            0.000000             NaN           0.000000                 NaN   
9            0.000000             NaN           0.000000                 NaN   
10           0.000000             NaN           0.000000                 NaN   
11           0.000000             NaN           0.000000                 NaN   
12           0.000000             NaN           0.000000                 NaN   
13           0.000000             NaN           0.000000                 NaN   
14           0.000000             NaN           0.000000                 NaN   
15           0.000000             NaN           0.000000                 NaN   
16           0.000000             NaN           0.000000                 NaN   
17           0.000000             NaN           0.000000                 NaN   
18           0.000000             NaN           0.000000                 NaN   
19           0.000000             NaN           0.000000                 NaN   
20           0.000000             NaN           0.000000                 NaN   
21           0.000000             NaN           0.000000                 NaN   
22           0.000000             NaN           0.010989                 NaN   
23           0.000000             NaN           0.000000                 NaN   
24           0.000000             NaN           0.000000                 NaN   
25           0.000000             NaN           0.000000                 NaN   
26           0.000000             NaN           0.000000                 NaN   
27           0.000000             NaN           0.000000                 NaN   
28           0.000000             NaN           0.000000                 NaN   
29           0.000000             NaN           0.000000                 NaN   
..                ...             ...                ...                 ...   
10           0.000000            0.00           0.000000                 0.0   
11           0.000000            0.00           0.000000                 0.0   
12           0.000000            0.00           0.000000                 0.0   
13           0.000000            0.00           0.000000                 0.0   
14           0.000000            0.00           0.000000                 0.0   
15           0.000000            0.00           0.000000                 0.0   
16           0.000000            0.00           0.010000                 0.0   
17           0.000000            0.00           0.000000                 0.0   
18           0.000000            0.00           0.000000                 0.0   
19           0.000000            0.00           0.000000                 0.0   
20           0.000000            0.00           0.000000                 0.0   
21           0.000000            0.00           0.000000                 0.0   
22           0.000000            0.00           0.000000                 0.0   
23           0.000000            0.00           0.000000                 0.0   
24           0.000000            0.00           0.000000                 0.0   
25           0.000000            0.00           0.000000                 0.0   
26           0.000000            0.00           0.000000                 0.0   
27           0.000000            0.01           0.000000                 0.0   
28  

In [162]:
total_group.set_index('Neighborhood', inplace=True)

In [163]:
total_group.fillna(0, inplace=True)

In [164]:
total_group.head()

Accessories Store  \
Neighborhood                                                            
Adelaide, King, Richmond_t                                        0.0   
Agincourt_t                                                       0.0   
Agincourt North, L'Amoreaux East, Milliken, Ste...                0.0   
Albion Gardens, Beaumond Heights, Humbergate, J...                0.0   
Alderwood, Long Branch_t                                          0.0   

                                                    Adult Boutique  \
Neighborhood                                                         
Adelaide, King, Richmond_t                                     0.0   
Agincourt_t                                                    0.0   
Agincourt North, L'Amoreaux East, Milliken, Ste...             0.0   
Albion Gardens, Beaumond Heights, Humbergate, J...             0.0   
Alderwood, Long Branch_t                                       0.0   

                                                    Afghan Restaurant  \
Neighborhood                                                            
Adelaide, King, Richmond_t                                        0.0   
Agincourt_t                                                       0.0   
Agincourt North, L'Amoreaux East, Milliken, Ste...                0.0   
Albion Gardens, Beaumond Heights, Humbergate, J...                0.0   
Alderwood, Long Branch_t                                          0.0   

                                                    African Restaurant  \
Neighborhood                                                             
Adelaide, King, Richmond_t                                         0.0   
Agincourt_t                                                        0.0   
Agincourt North, L'Amoreaux East, Milliken, Ste...                 0.0   
Albion Gardens, Beaumond Heights, Humbergate, J...                 0.0   
Alderwood, Long Branch_t                                           0.0   

                                                    Airport  \
Neighborhood                                                  
Adelaide, King, Richmond_t                              0.0   
Agincourt_t                                             0.0   
Agincourt North, L'Amoreaux East, Milliken, Ste...      0.0   
Albion Gardens, Beaumond Heights, Humbergate, J...      0.0   
Alderwood, Long Branch_t                                0.0   

                                                    Airport Food Court  \
Neighborhood                                                             
Adelaide, King, Richmond_t                                         0.0   
Agincourt_t                                                        0.0   
Agincourt North, L'Amoreaux East, Milliken, Ste...                 0.0   
Albion Gardens, Beaumond Heights, Humbergate, J...                 0.0   
Alderwood, Long Branch_t                                           0.0   

                                                    Airport Gate  \
Neighborhood                                                       
Adelaide, King, Richmond_t                                   0.0   
Agincourt_t                                                  0.0   
Agincourt North, L'Amoreaux East, Milliken, Ste...           0.0   
Albion Gardens, Beaumond Heights, Humbergate, J...           0.0   
Alderwood, Long Branch_t                                     0.0   

                                                    Airport Lounge  \
Neighborhood                                                         
Adelaide, King, Richmond_t                                     0.0   
Agincourt_t                                                    0.0   
Agincourt North, L'Amoreaux East, Milliken, Ste...             0.0   
Albion Gardens, Beaumond Heights, Humbergate, J...             0.0   
Alderwood, Long Branch_t                                       0.0   

                                                    Airport Service  \
Neighborhood      

In [166]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [168]:
# set number of clusters
kclusters = 10

#manhattan_grouped_clustering = manhattan_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(total_group)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([7, 7, 0, 1, 1, 1, 7, 1, 7, 7], dtype=int32)

In [169]:
# add clustering labels
total_group.insert(0, 'Cluster Labels', kmeans.labels_)



#### Cluster 1

In [170]:
total_group.loc[total_group['Cluster Labels'] == 0, total_group.columns[[1] + list(range(5, total_group.shape[1]))]]

Accessories Store  \
Neighborhood                                                            
Agincourt North, L'Amoreaux East, Milliken, Ste...                0.0   
Moore Park, Summerhill East_t                                     0.0   
Scarborough Village_t                                             0.0   

                                                    Airport  \
Neighborhood                                                  
Agincourt North, L'Amoreaux East, Milliken, Ste...      0.0   
Moore Park, Summerhill East_t                           0.0   
Scarborough Village_t                                   0.0   

                                                    Airport Food Court  \
Neighborhood                                                             
Agincourt North, L'Amoreaux East, Milliken, Ste...                 0.0   
Moore Park, Summerhill East_t                                      0.0   
Scarborough Village_t                                              0.0   

                                                    Airport Gate  \
Neighborhood                                                       
Agincourt North, L'Amoreaux East, Milliken, Ste...           0.0   
Moore Park, Summerhill East_t                                0.0   
Scarborough Village_t                                        0.0   

                                                    Airport Lounge  \
Neighborhood                                                         
Agincourt North, L'Amoreaux East, Milliken, Ste...             0.0   
Moore Park, Summerhill East_t                                  0.0   
Scarborough Village_t                                          0.0   

                                                    Airport Service  \
Neighborhood                                                          
Agincourt North, L'Amoreaux East, Milliken, Ste...              0.0   
Moore Park, Summerhill East_t                                   0.0   
Scarborough Village_t                                           0.0   

                                                    Airport Terminal  \
Neighborhood                                                           
Agincourt North, L'Amoreaux East, Milliken, Ste...               0.0   
Moore Park, Summerhill East_t                                    0.0   
Scarborough Village_t                                            0.0   

                                                    American Restaurant  \
Neighborhood                                                              
Agincourt North, L'Amoreaux East, Milliken, Ste...                  0.0   
Moore Park, Summerhill East_t                                       0.0   
Scarborough Village_t                                               0.0   

                                                    Animal Shelter  \
Neighborhood                                                         
Agincourt North, L'Amoreaux East, Milliken, Ste...             0.0   
Moore Park, Summerhill East_t                                  0.0   
Scarborough Village_t                                          0.0   

                                                    Antique Shop  ...  \
Neighborhood                                                      ...   
Agincourt North, L'Amoreaux East, Milliken, Ste...           0.0  ...   
Moore Park, Summerhill East_t                                0.0  ...   
Scarborough Village_t                                        0.0  ...   

                                                    Volleyball Court  \
Neighborhood                                                           
Agincourt North, L'Amoreaux East, Milliken, Ste...               0.0   
Moore Park, Summerhill East_t                                    0.0   
Scarborough Village_t                                            0.0   

                                                    Warehouse Store  \
Neighborhood                                                     

In [171]:
total_group.loc[total_group['Cluster Labels'] == 1, total_group.columns[[1] + list(range(5, total_group.shape[1]))]]

Accessories Store  \
Neighborhood                                                            
Albion Gardens, Beaumond Heights, Humbergate, J...                0.0   
Alderwood, Long Branch_t                                          0.0   
Bathurst Manor, Downsview North, Wilson Heights_t                 0.0   
Bedford Park, Lawrence Manor East_t                               0.0   
Bloordale Gardens, Eringate, Markland Wood, Old...                0.0   
Canada Post Gateway Processing Centre_t                           0.0   
Clarks Corners, Sullivan, Tam O'Shanter_t                         0.0   
Deer Park, Forest Hill SE, Rathnelly, South Hil...                0.0   
East Birchmount Park, Ionview, Kennedy Park_t                     0.0   
Glencairn_t                                                       0.0   
Guildwood, Morningside, West Hill_t                               0.0   
Humber Summit_t                                                   0.0   
Kingsview Village, Martin Grove Gardens, Richvi...                0.0   
L'Amoreaux West_t                                                 0.0   
Queen's Park_t                                                    0.0   
The Annex, North Midtown, Yorkville_t                             0.0   
Victoria Village_t                                                0.0   
Westmount_t                                                       0.0   
Willowdale West_t                                                 0.0   
Woburn_t                                                          0.0   
Woodbine Gardens, Parkview Hill_t                                 0.0   

                                                    Airport  \
Neighborhood                                                  
Albion Gardens, Beaumond Heights, Humbergate, J...      0.0   
Alderwood, Long Branch_t                                0.0   
Bathurst Manor, Downsview North, Wilson Heights_t       0.0   
Bedford Park, Lawrence Manor East_t                     0.0   
Bloordale Gardens, Eringate, Markland Wood, Old...      0.0   
Canada Post Gateway Processing Centre_t                 0.0   
Clarks Corners, Sullivan, Tam O'Shanter_t               0.0   
Deer Park, Forest Hill SE, Rathnelly, South Hil...      0.0   
East Birchmount Park, Ionview, Kennedy Park_t           0.0   
Glencairn_t                                             0.0   
Guildwood, Morningside, West Hill_t                     0.0   
Humber Summit_t                                         0.0   
Kingsview Village, Martin Grove Gardens, Richvi...      0.0   
L'Amoreaux West_t                                       0.0   
Queen's Park_t                                          0.0   
The Annex, North Midtown, Yorkville_t                   0.0   
Victoria Village_t                                      0.0   
Westmount_t                                             0.0   
Willowdale West_t                                       0.0   
Woburn_t                                                0.0   
Woodbine Gardens, Parkview Hill_t                       0.0   

                                                    Airport Food Court  \
Neighborhood                                                             
Albion Gardens, Beaumond Heights, Humbergate, J...                 0.0   
Alderwood, Long Branch_t                                           0.0   
Bathurst Manor, Downsview North, Wilson Heights_t                  0.0   
Bedford Park, Lawrence Manor East_t                                0.0   
Bloordale Gardens, Eringate, Markland Wood, Old...                 0.0   
Canada Post Gateway Processing Centre_t                            0.0   
Clarks Corners, Sullivan, Tam O'Shanter_t                          0.0   
Deer Park, Forest Hill SE, Rathnelly, South Hil...                 0.0   
East Birchmount Park, Ionview, Kennedy Park_t                      0.0   
Glencairn_t                                                        0.0   
Guildwood, Morningside, West Hill_

In [172]:
total_group.loc[total_group['Cluster Labels'] == 2, total_group.columns[[1] + list(range(5, total_group.shape[1]))]]

Accessories Store  \
Neighborhood                                                            
Cloverdale, Islington, Martin Grove, Princess G...                0.0   

                                                    Airport  \
Neighborhood                                                  
Cloverdale, Islington, Martin Grove, Princess G...      0.0   

                                                    Airport Food Court  \
Neighborhood                                                             
Cloverdale, Islington, Martin Grove, Princess G...                 0.0   

                                                    Airport Gate  \
Neighborhood                                                       
Cloverdale, Islington, Martin Grove, Princess G...           0.0   

                                                    Airport Lounge  \
Neighborhood                                                         
Cloverdale, Islington, Martin Grove, Princess G...             0.0   

                                                    Airport Service  \
Neighborhood                                                          
Cloverdale, Islington, Martin Grove, Princess G...              0.0   

                                                    Airport Terminal  \
Neighborhood                                                           
Cloverdale, Islington, Martin Grove, Princess G...               0.0   

                                                    American Restaurant  \
Neighborhood                                                              
Cloverdale, Islington, Martin Grove, Princess G...                  0.0   

                                                    Animal Shelter  \
Neighborhood                                                         
Cloverdale, Islington, Martin Grove, Princess G...             0.0   

                                                    Antique Shop  ...  \
Neighborhood                                                      ...   
Cloverdale, Islington, Martin Grove, Princess G...           0.0  ...   

                                                    Volleyball Court  \
Neighborhood                                                           
Cloverdale, Islington, Martin Grove, Princess G...               0.0   

                                                    Warehouse Store  \
Neighborhood                                                          
Cloverdale, Islington, Martin Grove, Princess G...              0.0   

                                                    Waterfront  \
Neighborhood                                                     
Cloverdale, Islington, Martin Grove, Princess G...         0.0   

                                                    Weight Loss Center  \
Neighborhood                                                             
Cloverdale, Islington, Martin Grove, Princess G...                 0.0   

                                                    Whisky Bar  Wine Bar  \
Neighborhood                                                               
Cloverdale, Islington, Martin Grove, Princess G...         0.0       0.0   

                                                    Wine Shop  Wings Joint  \
Neighborhood                                                                 
Cloverdale, Islington, Martin Grove, Princess G...        0.0          0.0   

                                                    Women's Store  Yoga Studio  
Neighborhood                                                                    
Cloverdale, Islington, Martin Grove, Princess G...            0.0          0.0  

[1 rows x 396 columns]

In [173]:
total_group.loc[total_group['Cluster Labels'] == 3, total_group.columns[[1] + list(range(5, total_group.shape[1]))]]

Accessories Store  Airport  Airport Food Court  \
Neighborhood                                                                 
Silver Hills, York Mills_t                0.0      0.0                 0.0   

                            Airport Gate  Airport Lounge  Airport Service  \
Neighborhood                                                                
Silver Hills, York Mills_t           0.0             0.0              0.0   

                            Airport Terminal  American Restaurant  \
Neighborhood                                                        
Silver Hills, York Mills_t               0.0                  0.0   

                            Animal Shelter  Antique Shop  ...  \
Neighborhood                                              ...   
Silver Hills, York Mills_t             0.0           0.0  ...   

                            Volleyball Court  Warehouse Store  Waterfront  \
Neighborhood                                                                
Silver Hills, York Mills_t               0.0              0.0         0.0   

                            Weight Loss Center  Whisky Bar  Wine Bar  \
Neighborhood                                                           
Silver Hills, York Mills_t                 0.0         0.0       0.0   

                            Wine Shop  Wings Joint  Women's Store  Yoga Studio  
Neighborhood                                                                    
Silver Hills, York Mills_t        0.0          0.0            0.0          0.0  

[1 rows x 396 columns]

In [174]:
total_group.loc[total_group['Cluster Labels'] == 4, total_group.columns[[1] + list(range(5, total_group.shape[1]))]]

Accessories Store  Airport  \
Neighborhood                                                                  
CFB Toronto, Downsview East_t                                  0.0      0.5   
Caledonia-Fairbanks_t                                          0.0      0.0   
East Toronto_t                                                 0.0      0.0   
Parkwoods_t                                                    0.0      0.0   
Rosedale_t                                                     0.0      0.0   
The Kingsway, Montgomery Road, Old Mill North_t                0.0      0.0   
Weston_t                                                       0.0      0.0   
York Mills West_t                                              0.0      0.0   

                                                 Airport Food Court  \
Neighborhood                                                          
CFB Toronto, Downsview East_t                                   0.0   
Caledonia-Fairbanks_t                                           0.0   
East Toronto_t                                                  0.0   
Parkwoods_t                                                     0.0   
Rosedale_t                                                      0.0   
The Kingsway, Montgomery Road, Old Mill North_t                 0.0   
Weston_t                                                        0.0   
York Mills West_t                                               0.0   

                                                 Airport Gate  Airport Lounge  \
Neighborhood                                                                    
CFB Toronto, Downsview East_t                             0.0             0.0   
Caledonia-Fairbanks_t                                     0.0             0.0   
East Toronto_t                                            0.0             0.0   
Parkwoods_t                                               0.0             0.0   
Rosedale_t                                                0.0             0.0   
The Kingsway, Montgomery Road, Old Mill North_t           0.0             0.0   
Weston_t                                                  0.0             0.0   
York Mills West_t                                         0.0             0.0   

                                                 Airport Service  \
Neighborhood                                                       
CFB Toronto, Downsview East_t                                0.0   
Caledonia-Fairbanks_t                                        0.0   
East Toronto_t                                               0.0   
Parkwoods_t                                                  0.0   
Rosedale_t                                                   0.0   
The Kingsway, Montgomery Road, Old Mill North_t              0.0   
Weston_t                                                     0.0   
York Mills West_t                                            0.0   

                                                 Airport Terminal  \
Neighborhood                                                        
CFB Toronto, Downsview East_t                                 0.0   
Caledonia-Fairbanks_t                                         0.0   
East Toronto_t                                                0.0   
Parkwoods_t                                                   0.0   
Rosedale_t                                                    0.0   
The Kingsway, Montgomery Road, Old Mill North_t               0.0   
Weston_t                                                      0.0   
York Mills West_t                                             0.0   

                                                 American Restaurant  \
Neighborhood                                                           
CFB Toronto, Downsview East_t                                    0.0   
Caledonia-Fairbanks_t                                            0.0   
East Toronto_t                                                   0.0   
Park

In [175]:
total_group.loc[total_group['Cluster Labels'] == 5, total_group.columns[[1] + list(range(5, total_group.shape[1]))]]

Accessories Store  Airport  Airport Food Court  \
Neighborhood                                                       
Rouge, Malvern_t                0.0      0.0                 0.0   

                  Airport Gate  Airport Lounge  Airport Service  \
Neighborhood                                                      
Rouge, Malvern_t           0.0             0.0              0.0   

                  Airport Terminal  American Restaurant  Animal Shelter  \
Neighborhood                                                              
Rouge, Malvern_t               0.0                  0.0             0.0   

                  Antique Shop  ...  Volleyball Court  Warehouse Store  \
Neighborhood                    ...                                      
Rouge, Malvern_t           0.0  ...               0.0              0.0   

                  Waterfront  Weight Loss Center  Whisky Bar  Wine Bar  \
Neighborhood                                                             
Rouge, Malvern_t         0.0                 0.0         0.0       0.0   

                  Wine Shop  Wings Joint  Women's Store  Yoga Studio  
Neighborhood                                                          
Rouge, Malvern_t        0.0          0.0            0.0          0.0  

[1 rows x 396 columns]

In [176]:
total_group.loc[total_group['Cluster Labels'] == 6, total_group.columns[[1] + list(range(5, total_group.shape[1]))]]

Accessories Store  Airport  Airport Food Court  Airport Gate  \
Neighborhood                                                                 
Roselawn_t                  0.0      0.0                 0.0           0.0   

              Airport Lounge  Airport Service  Airport Terminal  \
Neighborhood                                                      
Roselawn_t               0.0              0.0               0.0   

              American Restaurant  Animal Shelter  Antique Shop  ...  \
Neighborhood                                                     ...   
Roselawn_t                    0.0             0.0           0.0  ...   

              Volleyball Court  Warehouse Store  Waterfront  \
Neighborhood                                                  
Roselawn_t                 0.0              0.0         0.0   

              Weight Loss Center  Whisky Bar  Wine Bar  Wine Shop  \
Neighborhood                                                        
Roselawn_t                   0.0         0.0       0.0        0.0   

              Wings Joint  Women's Store  Yoga Studio  
Neighborhood                                           
Roselawn_t            0.0            0.0          0.0  

[1 rows x 396 columns]

In [177]:
total_group.loc[total_group['Cluster Labels'] == 7, total_group.columns[[1] + list(range(5, total_group.shape[1]))]]

Accessories Store  \
Neighborhood                                                            
Adelaide, King, Richmond_t                                   0.000000   
Agincourt_t                                                  0.000000   
Bayview Village_t                                            0.000000   
Berczy Park_t                                                0.000000   
Birch Cliff, Cliffside West_t                                0.000000   
Brockton, Exhibition Place, Parkdale Village_t               0.000000   
Business Reply Mail Processing Centre 969 Easte...           0.000000   
CN Tower, Bathurst Quay, Island airport, Harbou...           0.000000   
Cabbagetown, St. James Town_t                                0.000000   
Cedarbrae_t                                                  0.000000   
Central Bay Street_t                                         0.000000   
Chinatown, Grange Park, Kensington Market_t                  0.000000   
Christie_t                                                   0.000000   
Church and Wellesley_t                                       0.000000   
Clairlea, Golden Mile, Oakridge_t                            0.000000   
Cliffcrest, Cliffside, Scarborough Village West_t            0.000000   
Commerce Court, Victoria Hotel_t                             0.000000   
Davisville_t                                                 0.000000   
Davisville North_t                                           0.000000   
Del Ray, Keelesdale, Mount Dennis, Silverthorn_t             0.000000   
Design Exchange, Toronto Dominion Centre_t                   0.000000   
Don Mills North_t                                            0.000000   
Dorset Park, Scarborough Town Centre, Wexford H...           0.000000   
Dovercourt Village, Dufferin_t                               0.000000   
Downsview Northwest_t                                        0.000000   
Downsview West_t                                             0.000000   
Downsview, North Park, Upwood Park_t                         0.000000   
Fairview, Henry Farm, Oriole_t                               0.000000   
First Canadian Place, Underground city_t                     0.000000   
Flemingdon Park, Don Mills South_t                           0.000000   
...                                                               ...   
Flatiron_n                                                   0.000000   
Gramercy_n                                                   0.000000   
Greenwich Village_n                                          0.000000   
Hamilton Heights_n                                           0.000000   
Hudson Yards_n                                               0.000000   
Inwood_n                                                     0.000000   
Lenox Hill_n                                                 0.000000   
Lincoln Square_n                                             0.000000   
Little Italy_n                                               0.000000   
Lower East Side_n                                            0.000000   
Manhattan Valley_n                                           0.000000   
Manhattanville_n                                             0.000000   
Marble Hill_n                                                0.000000   
Midtown_n                                                    0.000000   
Midtown South_n                                              0.000000   
Morningside Heights_n                                        0.000000   
Murray Hill_n                                                0.000000   
Noho_n                                                       0.000000   
Roosevelt Island_n                                           0.000000   
Soho_n                                                       0.010000   
Stuyvesant Town_n                                            0.000000   
Sutton Place_n                                               0.000000   
Tribeca_n                            

In [178]:
total_group.loc[total_group['Cluster Labels'] == 8, total_group.columns[[1] + list(range(5, total_group.shape[1]))]]

Accessories Store  Airport  \
Neighborhood                                                  
The Junction North, Runnymede_t                0.0      0.0   

                                 Airport Food Court  Airport Gate  \
Neighborhood                                                        
The Junction North, Runnymede_t                 0.0           0.0   

                                 Airport Lounge  Airport Service  \
Neighborhood                                                       
The Junction North, Runnymede_t             0.0              0.0   

                                 Airport Terminal  American Restaurant  \
Neighborhood                                                             
The Junction North, Runnymede_t               0.0                  0.0   

                                 Animal Shelter  Antique Shop  ...  \
Neighborhood                                                   ...   
The Junction North, Runnymede_t             0.0           0.0  ...   

                                 Volleyball Court  Warehouse Store  \
Neighborhood                                                         
The Junction North, Runnymede_t               0.0              0.0   

                                 Waterfront  Weight Loss Center  Whisky Bar  \
Neighborhood                                                                  
The Junction North, Runnymede_t         0.0                 0.0         0.0   

                                 Wine Bar  Wine Shop  Wings Joint  \
Neighborhood                                                        
The Junction North, Runnymede_t       0.0        0.0          0.0   

                                 Women's Store  Yoga Studio  
Neighborhood                                                 
The Junction North, Runnymede_t            0.0          0.0  

[1 rows x 396 columns]

In [179]:
total_group.loc[total_group['Cluster Labels'] == 9, total_group.columns[[1] + list(range(5, total_group.shape[1]))]]

Accessories Store  Airport  Airport Food Court  \
Neighborhood                                                          
Downsview Central_t                0.0      0.0                 0.0   
Emery, Humberlea_t                 0.0      0.0                 0.0   

                     Airport Gate  Airport Lounge  Airport Service  \
Neighborhood                                                         
Downsview Central_t           0.0             0.0              0.0   
Emery, Humberlea_t            0.0             0.0              0.0   

                     Airport Terminal  American Restaurant  Animal Shelter  \
Neighborhood                                                                 
Downsview Central_t               0.0                  0.0             0.0   
Emery, Humberlea_t                0.0                  0.0             0.0   

                     Antique Shop  ...  Volleyball Court  Warehouse Store  \
Neighborhood                       ...                                      
Downsview Central_t           0.0  ...               0.0              0.0   
Emery, Humberlea_t            0.0  ...               0.0              0.0   

                     Waterfront  Weight Loss Center  Whisky Bar  Wine Bar  \
Neighborhood                                                                
Downsview Central_t         0.0                 0.0         0.0       0.0   
Emery, Humberlea_t          0.0                 0.0         0.0       0.0   

                     Wine Shop  Wings Joint  Women's Store  Yoga Studio  
Neighborhood                                                             
Downsview Central_t        0.0          0.0            0.0          0.0  
Emery, Humberlea_t         0.0          0.0            0.0          0.0  

[2 rows x 396 columns]

## Results and Discussion <a name="results"></a>

### Venue

There are 278 different kinds of venue in the 100 neibourhoods of Toronto.
There are 339 different kinds of venue in the 40 neibourhoods of Manhattan.
Toronto has more kinds of public services, such as hospital, Church and Beach.
Manhattan has more kinds of business.

### Restaurant tpye

There are 53 different kinds of Restaurants in Toronto.
There are 76 different kinds of Restaurants in Manhattan.
In Manhattan, you can choose differnet kinds of restaurants all over the world, including Afica, Tucky and Russia.

### Suberb tpye

In the K-MEAN Cluster, we can see all the neighborhoods in Manhattan are grouped in one cluster.
While neighborhoods in Tonronto have many differnt kinds of groups.

## Conclusion <a name="conclusion"></a>

Toronto and New York both are big city and share some same features, such as lots of different type of venues and business.

However, New York city has more diversity in culture and business. It has more different kinds of Resuarents as well.
All the suburbs in the test set in New York city have more Commonality for the reason that they are all in rural.